prov/rxm/src/rxm_conn.c

```diff
@@ -557,70 +563,126 @@ void rxm_cmap_process_reject(struct rxm_cmap *cmap,
 	}
 }
 
-int rxm_cmap_process_connreq(struct rxm_cmap *cmap, void *addr,
-			     struct rxm_cmap_handle **handle_ret,
-			     uint8_t *reject_reason)
+static int
+rxm_cmap_get_connreq_handle(struct rxm_cmap *cmap, void *addr,
+			    fi_addr_t *fi_addr, struct rxm_cmap_handle **handle)
 {
-	struct rxm_cmap_handle *handle;
-	int ret = 0, cmp;
-	fi_addr_t fi_addr = ofi_ip_av_get_fi_addr(cmap->av, addr);
+	int ret = 0;
 
+	*fi_addr = ofi_ip_av_get_fi_addr(cmap->av, addr);
 	ofi_straddr_dbg(cmap->av->prov, FI_LOG_EP_CTRL,
-			"Processing connreq from remote pep", addr);
+			"processing connreq from remote pep", addr);
 
-	if (fi_addr == FI_ADDR_NOTAVAIL)
-		handle = rxm_cmap_get_handle_peer(cmap, addr);
+	if (*fi_addr == FI_ADDR_NOTAVAIL)
+		*handle = rxm_cmap_get_handle_peer(cmap, addr);
 	else
-		handle = rxm_cmap_acquire_handle(cmap, fi_addr);
+		*handle = rxm_cmap_acquire_handle(cmap, *fi_addr);
 
-	if (!handle) {
-		if (fi_addr == FI_ADDR_NOTAVAIL)
-			ret = rxm_cmap_alloc_handle_peer(cmap, addr,
-							 RXM_CMAP_CONNREQ_RECV,
-							 &handle);
+	if (!(*handle)) {
+		if (*fi_addr == FI_ADDR_NOTAVAIL)
+			ret = rxm_cmap_alloc_handle_peer(cmap, addr, handle);
 		else
-			ret = rxm_cmap_alloc_handle(cmap, fi_addr,
-						    RXM_CMAP_CONNREQ_RECV,
-						    &handle);
-		if (ret)
-			goto unlock;
+			ret = rxm_cmap_alloc_handle(cmap, *fi_addr, handle);
+		if (!ret) {
+			if (cmap->simplex)
+				RXM_CM_UPDATE_RX_STATE(*handle,
+						       RXM_CMAP_CONNREQ_RECV);
+			else
+				RXM_CM_UPDATE_STATE(*handle,
+						    RXM_CMAP_CONNREQ_RECV);
+		}
```

Comment:
> I find the structure of this function in its original form hard to follow, and I think it's because of the duplicate fi_addr check. Maybe this would be easier?
> 
> ```
> if (fi_addr == FI_ADDR_NOTAVAIL) {
>     handle = rxm_cmap_get_handle_peer(cmap, addr);
>     if (!handle)
>         ret = rxm_cmap_alloc_handle_peer(cmap, addr, handle);
> } else {
>     handle = rxm_cmap_acquire_handle(cmap, fi_addr);
>     if (!handle)
>         ret = rxm_cmap_alloc_handle(cmap, fi_addr, handle);
> }
> ```

I agree, I will add a commit to restructure the original code before applying the commits required by this PR. Thanks.